R/ggridges.R

```diff
@@ -0,0 +1,270 @@
+#' Get data for ridge plots
+#'
+#' @param data dataframe, the data returned by `ggplot2::ggplot_build()`.
+#' @param na.rm boolean, from params
+#'
+#' @return dataframe containing plotting data 
+#'
+get_ridge_data <- function(data, na.rm) {
+  if (isTRUE(na.rm)) {
+    data <- data[stats::complete.cases(data[c("x", "ymin", "ymax")]), ]
+  }
+  
+  #if dataframe is empty there's nothing to draw
+  if (nrow(data) == 0) return(list())
+  
+  # remove all points that fall below the minimum height
+  data$ymax[data$height < data$min_height] <- NA
+  
+  # order data
+  data <- data[order(data$ymin, data$x), ]
+  
+  # remove missing points
+  missing_pos <- !stats::complete.cases(data[c("x", "ymin", "ymax")])
+  ids <- cumsum(missing_pos) + 1
+  data$group <- paste0(data$group, "-", ids)
+  data[!missing_pos, ]
+}
+
+
+#' Prepare plotting data for ggridges
+#' @param closed boolean, should the polygon be closed at bottom (TRUE for
+#'   geom_density_ridges2, FALSE for geom_density_ridges)
+prepare_ridge_chart <- function(data, prestats_data, layout, params, p, closed = FALSE, ...){
+  
+  d <- get_ridge_data(data, params$na.rm)
+  
+  # split data into separate groups
+  groups <- split(d, factor(d$group))
+  
+  # sort list so lowest ymin values are in the front (opposite of ggridges)
+  o <- order(
+    unlist(
+      lapply(
+        groups,
+        function(data) data$ymin[1]
+      )
+    ),
+    decreasing = FALSE
+  )
+  groups <- groups[o]
+  
+  # for each group create a density + vline + point as applicable
+  res <- lapply(
+    rev(groups),
+    function(x){
+      
+      draw_stuff <- split(x, x$datatype)
+      
+      # first draw the basic density ridge part
+      
+      stopifnot(!is.null(draw_stuff$ridgeline))
+      d2 <- d1 <- draw_stuff$ridgeline
+      if(!closed) d2$colour <- NA # no colour for density bottom line
+      
+      d1$y <- d1$ymax
+      d1$alpha <- 1 # don't use fill alpha for line alpha
+      ridges <- list(
+        to_basic(prefix_class(d2, "GeomDensity")),
+        to_basic(prefix_class(d1, "GeomLine"))
+      )
+      
+      if('vline' %in% names(draw_stuff)){
```

Comment:
Please be consistent with whitespace (I prefer more rather than less)

```suggestion
      if('vline' %in% names(draw_stuff)) {
```

Reply:
Done